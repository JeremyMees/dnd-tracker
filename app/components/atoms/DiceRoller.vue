<script setup lang="ts">
import { useToast } from '~/components/ui/toast/use-toast'

const emit = defineEmits<{ rolled: [amount: number] }>()

withDefaults(defineProps<{
  styled?: boolean
}>(), {
  styled: true,
})

const { toast } = useToast()
const { t } = useI18n()

const dices: Dice[] = ['d4', 'd6', 'd8', 'd10', 'd12', 'd20', 'd100']
const toRoll = ref<Record<Dice, number>>({ d4: 0, d6: 0, d8: 0, d10: 0, d12: 0, d20: 0, d100: 0 })

function calculateDiceRoll() {
  const rolled: Partial<Record<Dice, number[]>> = {}

  for (const [dice, count] of Object.entries(toRoll.value)) {
    if (count > 0) {
      rolled[dice as Dice] = rollDice(+dice.replace('d', ''), count)
    }
  }

  const keys = Object.keys(rolled)
  const values = Object.values(rolled)
  const amount = values.reduce((sum, arr) => sum + (arr?.reduce((a, b) => a + b, 0) || 0), 0)

  toast({
    title: t('components.diceRoll.rolled', { amount: values[0]?.length ?? 0, dice: keys.join(', ') }, keys.length),
    description: h('div', { class: 'flex flex-col gap-y-1' }, [
      ...Object.entries(rolled).map(([dice, numbers]) =>
        h('span', [
          h('span', { class: 'font-bold' }, `${dice}: `),
          h('span',
            numbers?.map((num, index) => [
              h('span', {
                class: {
                  'text-success': num === +(dice.replace('d', '')),
                  'text-destructive': num === 1,
                },
              }, num),
              index < numbers.length - 1 ? ', ' : '',
            ]).flat(),
          ),
        ]),
      ),
      Object.values(rolled).flat().length > 1 && h('span', { class: 'flex flex-row' }, [
        h('span', { class: 'font-bold mr-2' }, `${t('general.total')}:`),
        h('span', Object.values(rolled).flat().reduce((a, b) => a + b, 0)),
      ]),
    ]),
  })

  toRoll.value = { d4: 0, d6: 0, d8: 0, d10: 0, d12: 0, d20: 0, d100: 0 }
  emit('rolled', amount)
}
</script>

<template>
  <div :class="{ 'bg-background rounded-lg p-2 mb-4': styled }">
    <div class="flex items-center gap-x-2">
      <div
        v-for="dice in dices"
        :key="dice"
        class="flex flex-col items-center flex-1"
      >
        <button
          type="button"
          :aria-label="$t('general.more')"
          :disabled="toRoll[dice] >= 100"
          class="icon-btn-ghost"
          @click="toRoll[dice] = toRoll[dice] ? toRoll[dice] + 1 : 1"
        >
          <Icon
            name="tabler:caret-up"
            class="size-5"
          />
        </button>
        <div class="size-5">
          <NuxtImg
            v-tippy="dice.toUpperCase()"
            :src="`/${dice}.webp`"
            :alt="dice.toUpperCase()"
            sizes="sm:20px md:20px lg:20px"
            class="w-full h-full object-contain"
            format="webp"
            provider="imagekit"
          />
        </div>
        <button
          type="button"
          :aria-label="$t('general.less')"
          :disabled="toRoll[dice] <= 0"
          class="icon-btn-ghost"
          @click="toRoll[dice] = toRoll[dice] ? toRoll[dice] - 1 : 0"
        >
          <Icon
            name="tabler:caret-down"
            class="size-5"
          />
        </button>
        <span class="text-xs">
          {{ toRoll[dice] }}
        </span>
      </div>
    </div>
    <button
      type="button"
      :disabled="Object.values(toRoll).every((value) => value === 0)"
      class="btn-primary mt-4 w-full"
      @click="calculateDiceRoll"
    >
      {{ $t('actions.roll') }}
    </button>
  </div>
</template>
