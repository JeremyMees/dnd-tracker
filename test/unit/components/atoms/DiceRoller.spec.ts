import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, vi } from 'vitest'
import DiceRoller from '~/components/atoms/DiceRoller.vue'

interface Props { styled?: boolean }

const props: Props = { styled: true }

describe('DiceRoller', () => {
  it('Should match snapshot', async () => {
    const component = await mountSuspended(DiceRoller, { props })
    expect(component.html()).toMatchSnapshot()
  })

  it('Should render dice rollers correctly', async () => {
    const component = await mountSuspended(DiceRoller, { props })

    expect(component.html()).toContain('d4')
    expect(component.html()).toContain('d6')
    expect(component.html()).toContain('d8')
    expect(component.html()).toContain('d10')
    expect(component.html()).toContain('d12')
    expect(component.html()).toContain('d20')
    expect(component.html()).toContain('d100')
  })
})
