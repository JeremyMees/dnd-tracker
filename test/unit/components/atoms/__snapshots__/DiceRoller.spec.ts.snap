// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DiceRoller > Should match snapshot 1`] = `
"<div class="bg-background rounded-lg p-2 mb-4">
  <div class="flex items-center gap-x-2">
    <div class="flex flex-col items-center flex-1"><button type="button" aria-label="general.more" class="icon-btn-ghost"><span class="iconify i-tabler:caret-up size-5" aria-hidden="true"></span></button>
      <div class="size-5"><img alt="D4" data-nuxt-img="" sizes="(max-width: 768px) 20px, (max-width: 1024px) 20px, 20px" srcset="https://ik.imagekit.io/c2es1qasw/d4.webp?tr=w-20,f-webp,q-90 20w, https://ik.imagekit.io/c2es1qasw/d4.webp?tr=w-40,f-webp,q-90 40w" class="w-full h-full object-contain" src="https://ik.imagekit.io/c2es1qasw/d4.webp?tr=w-40,f-webp,q-90"></div><button type="button" aria-label="general.less" disabled="" class="icon-btn-ghost"><span class="iconify i-tabler:caret-down size-5" aria-hidden="true"></span></button><span class="text-xs">0</span>
    </div>
    <div class="flex flex-col items-center flex-1"><button type="button" aria-label="general.more" class="icon-btn-ghost"><span class="iconify i-tabler:caret-up size-5" aria-hidden="true"></span></button>
      <div class="size-5"><img alt="D6" data-nuxt-img="" sizes="(max-width: 768px) 20px, (max-width: 1024px) 20px, 20px" srcset="https://ik.imagekit.io/c2es1qasw/d6.webp?tr=w-20,f-webp,q-90 20w, https://ik.imagekit.io/c2es1qasw/d6.webp?tr=w-40,f-webp,q-90 40w" class="w-full h-full object-contain" src="https://ik.imagekit.io/c2es1qasw/d6.webp?tr=w-40,f-webp,q-90"></div><button type="button" aria-label="general.less" disabled="" class="icon-btn-ghost"><span class="iconify i-tabler:caret-down size-5" aria-hidden="true"></span></button><span class="text-xs">0</span>
    </div>
    <div class="flex flex-col items-center flex-1"><button type="button" aria-label="general.more" class="icon-btn-ghost"><span class="iconify i-tabler:caret-up size-5" aria-hidden="true"></span></button>
      <div class="size-5"><img alt="D8" data-nuxt-img="" sizes="(max-width: 768px) 20px, (max-width: 1024px) 20px, 20px" srcset="https://ik.imagekit.io/c2es1qasw/d8.webp?tr=w-20,f-webp,q-90 20w, https://ik.imagekit.io/c2es1qasw/d8.webp?tr=w-40,f-webp,q-90 40w" class="w-full h-full object-contain" src="https://ik.imagekit.io/c2es1qasw/d8.webp?tr=w-40,f-webp,q-90"></div><button type="button" aria-label="general.less" disabled="" class="icon-btn-ghost"><span class="iconify i-tabler:caret-down size-5" aria-hidden="true"></span></button><span class="text-xs">0</span>
    </div>
    <div class="flex flex-col items-center flex-1"><button type="button" aria-label="general.more" class="icon-btn-ghost"><span class="iconify i-tabler:caret-up size-5" aria-hidden="true"></span></button>
      <div class="size-5"><img alt="D10" data-nuxt-img="" sizes="(max-width: 768px) 20px, (max-width: 1024px) 20px, 20px" srcset="https://ik.imagekit.io/c2es1qasw/d10.webp?tr=w-20,f-webp,q-90 20w, https://ik.imagekit.io/c2es1qasw/d10.webp?tr=w-40,f-webp,q-90 40w" class="w-full h-full object-contain" src="https://ik.imagekit.io/c2es1qasw/d10.webp?tr=w-40,f-webp,q-90"></div><button type="button" aria-label="general.less" disabled="" class="icon-btn-ghost"><span class="iconify i-tabler:caret-down size-5" aria-hidden="true"></span></button><span class="text-xs">0</span>
    </div>
    <div class="flex flex-col items-center flex-1"><button type="button" aria-label="general.more" class="icon-btn-ghost"><span class="iconify i-tabler:caret-up size-5" aria-hidden="true"></span></button>
      <div class="size-5"><img alt="D12" data-nuxt-img="" sizes="(max-width: 768px) 20px, (max-width: 1024px) 20px, 20px" srcset="https://ik.imagekit.io/c2es1qasw/d12.webp?tr=w-20,f-webp,q-90 20w, https://ik.imagekit.io/c2es1qasw/d12.webp?tr=w-40,f-webp,q-90 40w" class="w-full h-full object-contain" src="https://ik.imagekit.io/c2es1qasw/d12.webp?tr=w-40,f-webp,q-90"></div><button type="button" aria-label="general.less" disabled="" class="icon-btn-ghost"><span class="iconify i-tabler:caret-down size-5" aria-hidden="true"></span></button><span class="text-xs">0</span>
    </div>
    <div class="flex flex-col items-center flex-1"><button type="button" aria-label="general.more" class="icon-btn-ghost"><span class="iconify i-tabler:caret-up size-5" aria-hidden="true"></span></button>
      <div class="size-5"><img alt="D20" data-nuxt-img="" sizes="(max-width: 768px) 20px, (max-width: 1024px) 20px, 20px" srcset="https://ik.imagekit.io/c2es1qasw/d20.webp?tr=w-20,f-webp,q-90 20w, https://ik.imagekit.io/c2es1qasw/d20.webp?tr=w-40,f-webp,q-90 40w" class="w-full h-full object-contain" src="https://ik.imagekit.io/c2es1qasw/d20.webp?tr=w-40,f-webp,q-90"></div><button type="button" aria-label="general.less" disabled="" class="icon-btn-ghost"><span class="iconify i-tabler:caret-down size-5" aria-hidden="true"></span></button><span class="text-xs">0</span>
    </div>
    <div class="flex flex-col items-center flex-1"><button type="button" aria-label="general.more" class="icon-btn-ghost"><span class="iconify i-tabler:caret-up size-5" aria-hidden="true"></span></button>
      <div class="size-5"><img alt="D100" data-nuxt-img="" sizes="(max-width: 768px) 20px, (max-width: 1024px) 20px, 20px" srcset="https://ik.imagekit.io/c2es1qasw/d100.webp?tr=w-20,f-webp,q-90 20w, https://ik.imagekit.io/c2es1qasw/d100.webp?tr=w-40,f-webp,q-90 40w" class="w-full h-full object-contain" src="https://ik.imagekit.io/c2es1qasw/d100.webp?tr=w-40,f-webp,q-90"></div><button type="button" aria-label="general.less" disabled="" class="icon-btn-ghost"><span class="iconify i-tabler:caret-down size-5" aria-hidden="true"></span></button><span class="text-xs">0</span>
    </div>
  </div><button type="button" disabled="" class="btn-primary mt-4 w-full">actions.roll</button>
</div>"
`;
